### PR TITLE
Nfa button component

### PR DESCRIPTION
**Pull request by MaksymKlets**
Nfa button component

### CODE REVIEW

File: src/app/faq-assistant/services/communication-faq-assistant.service.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import {Injectable} from '@angular/core';
+import {Observable} from 'rxjs';
+import {Subject} from 'rxjs/Subject';
+
+@Injectable()
+export class CommunicationFaqAssistantService {
+  subject = new Subject<any>();
```
Comment (eugen-tikhonov) [inline src/app/faq-assistant/services/communication-faq-assistant.service.ts]: This should be private property. And please, give more specific name. 
And the type of Subject not `any`, it is `void`

File: src/app/faq-assistant/services/communication-faq-assistant.service.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+  constructor() {
+  }
+
+  sendVisibilityState(state: boolean): void {
+    this.subject.next(state);
+  }
+
+  receiveVisibilityState(): Observable<any> {
```
Comment (eugen-tikhonov) [inline src/app/faq-assistant/services/communication-faq-assistant.service.ts]: The type of Observable - `void`

File: src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    this.subscription = this.communicationFaqAssistantService.receiveVisibilityState().subscribe(stateContainer => {
+      this.visibilityContainer = stateContainer;
+    });
+  }
+
+  ngOnInit() {
+    this.getTitleList();
+  }
+
+  createComponent(answer: any, params: any): void {
```
Comment (eugen-tikhonov) [inline src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts]: This should be private method.

File: src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    setTimeout(() => {
+      this.finalAnswerContent.clear();
+      const factory = this.resolver.resolveComponentFactory(answer);
+      this.componentRef = this.finalAnswerContent.createComponent(factory);
+      this.componentRef.instance.data = params;
+    });
+  }
+
+  getTitleList(): void {
```
Comment (eugen-tikhonov) [inline src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts]: This should be private method.

File: src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    this.titleList = this.data;
+    this.itemManager.setItem(this.data);
+  }
+
+  clearDataComponent(): void {
```
Comment (eugen-tikhonov) [inline src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts]: This should be private method.

File: src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  @Input() params: any;
+  @ViewChild('finalAnswer', {read: ViewContainerRef}) finalAnswerContent: ViewContainerRef;
+
+  constructor(
+    private itemManager: FaqItemListService,
+    private viewContainerRef: ViewContainerRef,
+    private resolver: ComponentFactoryResolver,
+    private communicationFaqAssistantService: CommunicationFaqAssistantService
+  ) {
+    this.subscription = this.communicationFaqAssistantService.receiveVisibilityState().subscribe(stateContainer => {
```
Comment (eugen-tikhonov) [inline src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts]: Usually the best part for observables it is ngOnInit() {}

File: src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+  titleList: FaqItemContentInterface[];
+  items: FaqItemContentInterface[][];
+  showFinish = false;
+  lastStep: boolean;
+  showBack = false;
+  answer: string;
+  componentRef: any;
+  subscription: Subscription;
+  visibilityContainer = true;
```
Comment (eugen-tikhonov) [inline src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts]: I think default value should be false.

File: src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+export class FaqAssistantComponent implements OnInit {
+
+  titleList: FaqItemContentInterface[];
+  items: FaqItemContentInterface[][];
+  showFinish = false;
+  lastStep: boolean;
+  showBack = false;
+  answer: string;
+  componentRef: any;
+  subscription: Subscription;
```
Comment (eugen-tikhonov) [inline src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts]: Please, add more specific name.

File: src/app/faq-assistant/components/faq-assistant-button/faq-assistant-button.component.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+@Component({
+  selector: 'app-faq-assistant-button',
+  templateUrl: './faq-assistant-button.component.html',
+  styleUrls: ['./faq-assistant-button.component.scss'],
+  encapsulation: ViewEncapsulation.None
+})
+export class FaqAssistantButtonComponent {
+
+  visibilityContainer = true;
```
Comment (eugen-tikhonov) [inline src/app/faq-assistant/components/faq-assistant-button/faq-assistant-button.component.ts]: Prefer to use 'is', 'has' prefix for states. 
Moreover sound like you want to send some container, but not state. More preferable to call prop like `isContainerVisible`. 

Also you can use only one prop in our component, without `visibilityCloseButton`

**Review by eugen-tikhonov — CHANGES_REQUESTED** [review]
Need to do some rework.

File: src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts (hunk 3 above)
Comment (eugen-tikhonov) [reply src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts]: Also please add answer type.

File: src/app/faq-assistant/services/communication-faq-assistant.service.ts (hunk 1 above)
Comment (MaksymKlets) [reply src/app/faq-assistant/services/communication-faq-assistant.service.ts]: type void is not suitable for this property

File: src/app/faq-assistant/services/communication-faq-assistant.service.ts (hunk 2 above)
Comment (MaksymKlets) [reply src/app/faq-assistant/services/communication-faq-assistant.service.ts]: was applied

File: src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts (hunk 3 above)
Comment (MaksymKlets) [reply src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts]: was applied except setting up type for answer argument

File: src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts (hunk 4 above)
Comment (MaksymKlets) [reply src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts]: was applied

File: src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts (hunk 5 above)
Comment (MaksymKlets) [reply src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts]: was applied

File: src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts (hunk 6 above)
Comment (MaksymKlets) [reply src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts]: was applied

File: src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts (hunk 7 above)
Comment (MaksymKlets) [reply src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts]: was applied

File: src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts (hunk 8 above)
Comment (MaksymKlets) [reply src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts]: was applied

File: src/app/faq-assistant/components/faq-assistant-button/faq-assistant-button.component.ts (hunk 9 above)
Comment (MaksymKlets) [reply src/app/faq-assistant/components/faq-assistant-button/faq-assistant-button.component.ts]: was applied

File: src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+@Component({
+  selector: 'app-faq-assistant',
+  templateUrl: './faq-assistant.component.html',
+  styleUrls: ['./faq-assistant.component.scss'],
+  encapsulation: ViewEncapsulation.None
+})
+export class FaqAssistantComponent implements OnInit {
+
+  titleList: object;
+  items: Array<object>;
```
Comment (eugen-tikhonov) [inline src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts]: More preferable to write object[];

File: src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+  showFinish = false;
+  lastStep: boolean;
+  showBack = false;
+  answer: string;
+  componentRef: any;
+  stateFaqContainer: Subscription;
+  isContainerVisible = false;
+  indexConfigList: Array<string>;
```
Comment (eugen-tikhonov) [inline src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts]: The same as above.

File: src/app/faq-assistant/services/communication-faq-assistant.service.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
+import {Injectable} from '@angular/core';
+import {Observable} from 'rxjs';
+import {Subject} from 'rxjs/Subject';
+
+@Injectable()
+export class CommunicationFaqAssistantService {
+  private visibilityContainer = new Subject<any>();
```
Comment (eugen-tikhonov) [inline src/app/faq-assistant/services/communication-faq-assistant.service.ts]: The type of Subject is `boolean`

File: src/app/faq-assistant/services/communication-faq-assistant.service.ts [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+  constructor() {
+  }
+
+  sendVisibilityState(state: boolean): void {
+    this.visibilityContainer.next(state);
+  }
+
+  receiveVisibilityState(): Observable<any> {
```
Comment (eugen-tikhonov) [inline src/app/faq-assistant/services/communication-faq-assistant.service.ts]: The same as above.

File: src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+
+@Component({
+  selector: 'app-faq-assistant',
+  templateUrl: './faq-assistant.component.html',
+  styleUrls: ['./faq-assistant.component.scss'],
+  encapsulation: ViewEncapsulation.None
+})
+export class FaqAssistantComponent implements OnInit {
+
+  titleList: object;
```
Comment (eugen-tikhonov) [inline src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts]: Can we use more specific interface than `object`?

**Review by eugen-tikhonov — CHANGES_REQUESTED** [review]
Please, add some changes

File: src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts (hunk 10 above)
Comment (MaksymKlets) [reply src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts]: was applied

File: src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts (hunk 11 above)
Comment (MaksymKlets) [reply src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts]: was applied

File: src/app/faq-assistant/services/communication-faq-assistant.service.ts (hunk 12 above)
Comment (MaksymKlets) [reply src/app/faq-assistant/services/communication-faq-assistant.service.ts]: was applied

File: src/app/faq-assistant/services/communication-faq-assistant.service.ts (hunk 13 above)
Comment (MaksymKlets) [reply src/app/faq-assistant/services/communication-faq-assistant.service.ts]: was applied

File: src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts (hunk 14 above)
Comment (MaksymKlets) [reply src/app/faq-assistant/components/faq-assistant/faq-assistant.component.ts]: Sure, I will use this interface: FaqItemContentInterface<FaqObject>

**Pull request merged**